File: lib/features/library/data/app_events.dart
import 'package:drift/drift.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:uuid/uuid.dart';

import '../../../core/db/database.dart';
import '../../../core/db/providers.dart';

/// Persistent, user-facing problem inbox. Background failures (import,
/// subtitle download, metadata, auth) are logged here so they survive past a
/// transient SnackBar and across restarts. Only error/warning — never success.
class AppEventSink {
  AppEventSink(this._db, {Uuid? uuid}) : _uuid = uuid ?? const Uuid();

  final TonariDatabase _db;
  final Uuid _uuid;

  static const _retain = 200;

  Future<void> log({
    required String category,
    String severity = 'error',
    required String title,
    String detail = '',
    String? productId,
    String? workTitle,
    String? sourceName,
    String? actionKey,
  }) async {
    final now = DateTime.now();
    // Dedup: same (category, productId, title) bumps count instead of spamming
    // a new row — 115 rate limiting fires the same error in bursts.
    final existing =
        await (_db.select(_db.appEvents)
              ..where(
                (e) =>
                    e.category.equals(category) &
                    e.title.equals(title) &
                    (productId == null
                        ? e.productId.isNull()
                        : e.productId.equals(productId)),
              )
              ..orderBy([(e) => OrderingTerm.desc(e.lastAt)])
              ..limit(1))
            .getSingleOrNull();
    if (existing != null) {
      await (_db.update(_db.appEvents)..where((e) => e.id.equals(existing.id)))
          .write(
            AppEventsCompanion(
              detail: Value(detail),
              lastAt: Value(now),
              count: Value(existing.count + 1),
              read: const Value(false),
            ),
          );
      return;
    }
    await _db
        .into(_db.appEvents)
        .insert(
          AppEventsCompanion.insert(
            id: _uuid.v4(),
            createdAt: now,
            lastAt: now,
            category: category,
            severity: severity,
            title: title,
            detail: Value(detail),
            productId: Value(productId),
            workTitle: Value(workTitle),
            sourceName: Value(sourceName),
            actionKey: Value(actionKey),
          ),
        );
    await _trim();
  }

  Future<void> markAllRead() async {
    await (_db.update(_db.appEvents)..where((e) => e.read.equals(false))).write(
      const AppEventsCompanion(read: Value(true)),
    );
  }

  Future<void> dismiss(String id) async {
    await (_db.delete(_db.appEvents)..where((e) => e.id.equals(id))).go();
  }

  Future<void> clear() async {
    await _db.delete(_db.appEvents).go();
  }

  Future<void> _trim() async {
    final keep = _db.selectOnly(_db.appEvents)
      ..addColumns([_db.appEvents.id])
      ..orderBy([OrderingTerm.desc(_db.appEvents.lastAt)])
      ..limit(_retain);
    await (_db.delete(
      _db.appEvents,
    )..where((e) => e.id.isNotInQuery(keep))).go();
  }
}

final appEventSinkProvider = Provider<AppEventSink>((ref) {
  return AppEventSink(ref.watch(databaseProvider));
});

final appEventsProvider = StreamProvider<List<AppEvent>>((ref) {
  final db = ref.watch(databaseProvider);
  return (db.select(db.appEvents)
        ..orderBy([(e) => OrderingTerm.desc(e.lastAt)]))
      .watch();
});

final unreadEventCountProvider = StreamProvider<int>((ref) {
  final db = ref.watch(databaseProvider);
  final query = db.selectOnly(db.appEvents)
    ..addColumns([db.appEvents.id.count()])
    ..where(db.appEvents.read.equals(false));
  return query.watchSingle().map(
    (row) => row.read(db.appEvents.id.count()) ?? 0,
  );
});

File: native/TonariCore/Sources/TonariCore/Events/AppEvents.swift
import Foundation
import GRDB

/// The message inbox: background failures (imports, metadata, 115 auth) and
/// the outcome of long jobs the user walked away from, like backups. Kept
/// across launches so they outlive the alert that first reported them.
public enum AppEvents {
    static let retained = 200

    public enum Severity: String, Sendable {
        case error, warning, info
    }

    /// A shortcut the inbox offers next to the message.
    public enum Action: String, Sendable {
        /// Retry enrichment of works that used up their attempts.
        case enrich
        /// Sign in to 115 again.
        case reauth
        /// Sign in to DLsite again.
        case dlsiteLogin
    }

    public static func all(_ db: Database) throws -> [AppEvent] {
        try AppEvent.order(Column("last_at").desc).fetchAll(db)
    }

    public static func unreadCount(_ db: Database) throws -> Int {
        try AppEvent.filter(Column("read") == false).fetchCount(db)
    }
}

extension AppDatabase {
    /// A repeat of the same category, title and work bumps the earlier entry
    /// instead of adding a row: 115 rate limiting fails in bursts.
    public func logEvent(
        category: String,
        severity: AppEvents.Severity = .error,
        title: String,
        detail: String = "",
        productId: String? = nil,
        workTitle: String? = nil,
        sourceName: String? = nil,
        action: AppEvents.Action? = nil,
        at date: Date = .now
    ) throws {
        try writer.write { db in
            let existing = try AppEvent
                .filter(Column("category") == category && Column("title") == title)
                .filter(sql: "product_id IS ?", arguments: [productId])
                .order(Column("last_at").desc)
                .fetchOne(db)
            if var event = existing {
                event.detail = detail
                event.lastAt = date
                event.count += 1
                event.read = false
                try event.update(db)
                return
            }
            try AppEvent(
                id: UUID().uuidString.lowercased(), createdAt: date, lastAt: date,
                category: category, severity: severity.rawValue, title: title, detail: detail,
                productId: productId, workTitle: workTitle, sourceName: sourceName, actionKey: action?.rawValue,
                count: 1, read: false
            ).insert(db)
            try db.execute(
                sql: "DELETE FROM app_events WHERE id NOT IN (SELECT id FROM app_events ORDER BY last_at DESC LIMIT ?)",
                arguments: [AppEvents.retained]
            )
        }
    }

    public func markEventsRead() throws {
        try writer.write { try $0.execute(sql: "UPDATE app_events SET read = 1 WHERE read = 0") }
    }

    public func dismissEvent(_ id: String) throws {
        _ = try writer.write { try AppEvent.deleteOne($0, key: id) }
    }

    public func clearEvents() throws {
        _ = try writer.write { try AppEvent.deleteAll($0) }
    }
}
